llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp

```diff
@@ -1467,97 +1458,116 @@ bool RISCVVLOptimizer::isCandidate(const MachineInstr &MI) const {
-        assert(RISCV::VRRegClass.hasSubClassEq(RC) &&
-               "Expect LMUL 1 register class for vector as scalar operands!");
-        LLVM_DEBUG(dbgs() << "    Use this operand as a scalar operand\n");
-        continue;
-      }
-
-      if (mayReadPastVL(UserMI)) {
-        LLVM_DEBUG(dbgs() << "    Abort due to used by unsafe instruction\n");
-        CanReduceVL = false;
-        break;
-      }
-
-      // Tied operands might pass through.
-      if (UserOp.isTied()) {
-        LLVM_DEBUG(dbgs() << "    Abort due to user use it as tied operand\n");
-        CanReduceVL = false;
-        break;
-      }
-
-      const MCInstrDesc &Desc = UserMI.getDesc();
-      if (!RISCVII::hasVLOp(Desc.TSFlags) || !RISCVII::hasSEWOp(Desc.TSFlags)) {
-        LLVM_DEBUG(dbgs() << "    Abort due to lack of VL or SEW, assume that"
-                             " use VLMAX.\n");
-        CanReduceVL = false;
-        break;
-      }
-
-      unsigned VLOpNum = RISCVII::getVLOpNum(Desc);
-      const MachineOperand &VLOp = UserMI.getOperand(VLOpNum);
-      // Looking for a register VL that isn't X0.
-      if (!VLOp.isReg() || VLOp.getReg() == RISCV::X0) {
-        LLVM_DEBUG(dbgs() << "    Abort due to user use X0 as VL.\n");
-        CanReduceVL = false;
-        break;
-      }
-
-      if (!CommonVL) {
-        CommonVL = VLOp.getReg();
-      } else if (*CommonVL != VLOp.getReg()) {
-        LLVM_DEBUG(dbgs() << "    Abort due to users have different VL!\n");
-        CanReduceVL = false;
-        break;
-      }
-
-      // The SEW and LMUL of destination and source registers need to match.
-
-      // If the produced Dest is not a vector register, then it has no EEW or
-      // EMUL, so there is no need to check that producer and consumer LMUL and
-      // SEW match. We've already checked above that UserOp is a vector
-      // register.
-      if (!isVectorRegClass(MI.getOperand(0).getReg(), MRI)) {
-        LLVM_DEBUG(dbgs() << "    Abort due to register class mismatch between "
-                             "USE and DEF\n");
-        continue;
-      }
-
-      OperandInfo ConsumerInfo = getOperandInfo(UserMI, UserOp, MRI);
-      OperandInfo ProducerInfo = getOperandInfo(MI, MI.getOperand(0), MRI);
-      if (ConsumerInfo.isUnknown() || ProducerInfo.isUnknown() ||
-          !OperandInfo::EMULAndEEWAreEqual(ConsumerInfo, ProducerInfo)) {
-        LLVM_DEBUG(dbgs() << "    Abort due to incompatible or unknown "
-                             "information for EMUL or EEW.\n");
-        LLVM_DEBUG(dbgs() << "      ConsumerInfo is: " << ConsumerInfo << "\n");
-        LLVM_DEBUG(dbgs() << "      ProducerInfo is: " << ProducerInfo << "\n");
-        CanReduceVL = false;
-        break;
-      }
+    if (isVectorRegClass(MI.getOperand(0).getReg(), MRI))
+      checkUsers(CommonVL, CanReduceVL, MI);
+    else {
+      CommonVL = MI.getOperand(RISCVII::getVLOpNum(MI.getDesc())).getReg();
```

Comment:
Won't this change the VL to its own VL? What's the point?

Reply:
I removed it.